### ISSUE TITLE
Mac build issues 

### ISSUE THREAD

**Issue by artisteacher** [opentoonz/opentoonz#2292]
### Issue Summary

Following the build instructions for MacOS either only partially works or results in a crash on open. I’m wondering if the instructions need to be different for High Sierra? Any assistance would be appreciated!
### Screenshot or Video Reference

<img width="300" alt="1a44c0b1-65ca-4f22-a2d2-1b9c195f0167" src="https://user-images.githubusercontent.com/21159116/45524428-77f31280-b79c-11e8-9140-83814ae5399d.png">

### Steps to Reproduce

Following the step between 2 & 3 to add the stuff folder causes problems when trying to open the resulting build. On open, there is a warning notification behind the startup screen that says the menubar settings template is missing (the warning is barely visible & has to be dragged into view). At that point OT needs to force quit. Using the stuff folder from a normal OT install allows the build to open. However the mypaint brush icons are missing (except for the no brush icon) and changing to the vector tab in the style editor always causes an immediate crash. 
### Expected Results

For OT to open if the stuff directory is created, for the icons to all be visible, and for no crashing with the vector tab. 
### Actual Results

If there is no OpenToonz folder in the Applications directory, copying the stuff folder fails. If there is an existing stuff folder in Applications, then the terminal command places the stuff folder in the 1.2 stuff folder. Either way, due to the path issues, OT can’t open completely. 

 If there is an existing stuff folder and terminal command is changed for the source to be stuff/ then the contents of the stuff directory appear to copy correctly. However in that case, there is a silent crash on open.

When using an existing stuff folder, OT opens but continues to have path issues with mypaint brush icons & crashes when changing to the vector tab. The vector tab crash does occur periodically anyways, but this is every time. This crash seems to stem from CustomStyleManager::StyleLoaderTask::run()
### System Information
* **OpenToonz Version**:
* **Operating System**: 10.13.6
* **CPU**: Intel Core i7
* **Memory**: 16GB
* **Graphics Card**: Radeon Pro 460 & Intel HD Graphics 530
* **Graphics Tablet**:

**Comment by artisteacher** [comment]
Now that I've updated the OS to 10.14, I'm still struggling to get a completely successful build on MacOS following the current build instructions. I updated everything that could be updated (including QT) and there are definitely still path issues. If I try to create the stuff directory when building, I get a crash before the loading window finishes:
<img width="792" alt="screen shot 2018-10-12 at 7 54 34 pm" src="https://user-images.githubusercontent.com/21159116/46898634-dd4c2900-ce58-11e8-9d5c-3a6ebc6853d4.png">

Using an old stuff folder I can open my build of OT mostly successfully. Again any time I go to the vector tab, there's always an instant crash - no matter what level type or brush style is active. I tried renaming the custom style folder as a test and that actually allows me to access the vector tab without issue. So whatever is causing the vector tab crash (#2303) - seems to be due in some way to the custom/trail styles. I'll look into that further and report in the other issue. 

On a related note, both vector and texture icons display without issue, while the mypaint icons are still all missing - except for the "no brush" icon.

Is there anyone building successfully on Mac (without these issues) who has any advice?

**Comment by artisteacher** [comment]
Making the build & new stuff folder portable allows the app to open.

**Comment by artisteacher** [comment]
The portable version of my build continues to have an instant crash switching to the vector tab & the mypaint brush icons do not display. These are both not issues in a portable version of the released app.

**Comment by RodneyBaker** [comment]
> These are both not issues in a portable version of the released app.

I will guess that those areas might have references in the Windows registry that didn't make the transition well to the Mac port.   Just a wild guess though.

**Comment by yehjaemin** [comment]
I ran into the exact same startup issue. I solved it by copying the opentoonz/stuff folder to Applications/OpenToonz and renaming it as OpenToonz_stuff.

For your other build issue, if your specific problem is with crashing due to x86_64 system architecture problems, then #2499 introduces a change to one of the .cmake files that should make building on macs crash-free again.

Hope this helps!

**Comment by artisteacher** [comment]
@yehjaemin - thanks, I finally got that working. This does fix the saving crash I was getting when building myself.  The mypaint brush icons are still missing and switching to the vector tab continues to cause a crash.

Do you mean manually copying the stuff folder through Finder? That does work for me. For whatever reason, using:
> sudo cp -r stuff /Applications/OpenToonz/OpenToonz_stuff

in Terminal, still causes a crash on trying to open. Could it be a permissions issue? If I inspect the stuff folder copied through terminal, the permissions are set to read only. Is there anything to add to the copy command that will keep the read and write permissions?

**Comment by manongjohn** [comment]
If the permissions are correct on the "stuff" folder, make it `cp -pr stuff `. The `-p` will preserve permissions.

It sounds like the default `umask` setting has changed which would explain why your files created via Terminal are marked as readonly. Most unix/unix-like systems have a umask of 002 or 022, but yours must be set to at least 222 in order to only have readonly. Not sure if that is a newer macOS version setting.

One other way to fix that is to type `umask 022` before the cp command to ensure proper permissions are granted on new files. 022 will give owner read/write and everyone else just read.

**Comment by artisteacher** [comment]
Using `cp -pr stuff`  works! Thanks @manongjohn. Perhaps @yehjaemin can add that change with the other updates to the build instructions in #2497. 

I think the only 2 issues that remain with building on macOS are the missing mypaint brush icons and crashing when switching to the vector tab. Is anyone else building on mac having those problems?

**Comment by yehjaemin** [comment]
I added the changes to my existing edits in PR #2497.

**Comment by RodneyBaker** [comment]
@artisteacher 
The two issues remaining in this report are:
- [ ] Missing mypaint brush icons
- [ ] Crash when switching to the vector tab

Are these issues still valid?

**Comment by artisteacher** [comment]
Yes, they are still valid. Saving a scene also seems to cause a crash.

**Comment by artisteacher** [comment]
The saving issue seems to stem from toonz/sources/image/png/tiio_png.cpp:
<img width="550" alt="Screen Shot 2019-12-08 at 4 40 37 PM" src="https://user-images.githubusercontent.com/21159116/70396727-9664af80-19d9-11ea-9355-a9a5facc9494.png">

**Comment by artisteacher** [comment]
Last Ic checked, I didn't get a crash from switching to the vector tab.

@MrBlockCat - are you noticing any issues when you build? Do your mypaint brush icons show up? Can you reliably save without crashing?

**Comment by MrBlockCat** [comment]
I'll be able to test that in a few hours, I'll let you know!

**Comment by MrBlockCat** [comment]
Apart from the weird arrow thing I've been having, I'm not seeing any of the problems you mentioned which is good! I haven't saved a ton so if it doesn't always happen that's hard to 100% verify, but I did save without crashing, and the other stuff definitely seems fixed.

Out of curiosity, though, are you seeing the grey double arrow or the black one when you build @artisteacher? It would be nice if that could be narrowed down and maybe fixed! If it is because of the new QT version, we'll want to update QT for everything at some point to fix a few things like #1445 and I'm sure others, so it would be nice to fix everything related to QT updates before we do update.

**Comment by artisteacher** [comment]
Interesting. I updated QT - and everything installed with homebrew. I still can get a crash when saving but it's not consistent. It is easy to trigger if I open an existing scene directly after starting opentoonz and try to save without making any changes. The raster brush icons still aren't visible. I wonder if you have something different than I do in your build environment? What are you using for you QT path?

I do get the grey double arrow - it's only the slightest bit visible on hover. Without hovering, it blends in completely. It really does need to be black...

**Comment by MrBlockCat** [comment]
Here's what I've got in opentoonz/toonz/sources/CMakeLists.txt: <img width="720" alt="Screen Shot 2020-03-25 at 11 39 19 PM" src="https://user-images.githubusercontent.com/37186692/77607829-e7010980-6ef1-11ea-8e46-b3c3d1d8d047.png">

I'm building using Xcode 11.4 (not using Terminal), which may be different? I'm also using macOS Catalina, 10.15.4. Hope something there helps!

**Comment by MrBlockCat** [comment]
I did actually just notice what may very well be another bug only when building myself: On a recent nightly, in the Stage Schematic window, I can right click the camera and open the Camera Settings and close them just fine. However, on a manual build on macOS, I can do all those things, but when I close the Camera Settings window, OT crashes.

**Comment by artisteacher** [comment]
I've been changing line 161 of CMakeLists.txt instead of line 162. I guess I'll see if that makes any difference in combination with using Xcode.

@MrBlockCat - As far as the typo goes, I think we could likely fix it without causing issues. It might be best to wait & do it whenever we are able to update the qt version in the cmake file.

I'm not getting a crash from closing the camera settings window (building with Terminal). Once I have the Xcode build, I'll see if that makes a difference.

**Comment by turtletooth** [comment]
Speaking of mac builds. 
@artisteacher If I tweak the code for stop motion to work with macs, could you test it and help me get it working?

**Comment by artisteacher** [comment]
@turtletooth - you are the best! I would definitely love to help with that! I'll need to pick up a camera from school - we're working remotely for at least another month. I might be able to pick it up in the next few days, but I'll let you know when I have the camera.

**Comment by MrBlockCat** [comment]
@artisteacher Ah! I’m not the only one. I did that too the first time I tried to build with Xcode, and it wouldn’t even compile at all. Maybe the line should say 162 instead of 160 (and 32 bit equivalent) in the build directions to help people with that, it’s a little confusing...

**Comment by manongjohn** [comment]
If you look at the original script, 162 uses the environment variable defined by 161.
I have not had any issues building when changing 161.

**Comment by MrBlockCat** [comment]
@manongjohn Have you had no issues building when only changing 161, or also 162? If only 161, that’s interesting since I believe that was what caused my build issues when using Xcode... That said, I believe I can’t build using Terminal at all, so I don’t know if there’s something deeper going on there.

**Comment by manongjohn** [comment]
@MrBlockCat Those directions have always worked for me regardless if building command line or with Xcode.

Line 161 sets up the QT_PATH environment variable which interestingly enough contains the `/lib` path already. The original Line 162 sets up QT_LIB_PATH to search under QT_PATH which was established in Line 161.  Changing 161 is enough, assuming the path ends with `/lib` as it does now.  

You just bypassed 161 and did it right in 162. Problem with that is if QT_PATH is ever needed in the future for some other stuff, your variable is pointing to something that doesn't exist on your system.  Partly surprised that it doesn't cause you problems building with it being wrong.

FYI, I am the one who came up with the Xcode build instructions that I use on my macOS VMs that I use for testing some macOS related issues as well as testing my OTX macOS builds.  I recently configured a macOS VM for Catalina to test some builds against it if I need it and had no problems getting Xcode to build.

**Comment by artisteacher** [comment]
Alright, running through Xcode, I do get a crash closing out of the camera settings window.
<img width="760" alt="Screen Shot 2020-04-05 at 11 30 42 PM" src="https://user-images.githubusercontent.com/21159116/78520708-80070e80-7795-11ea-9c84-456a6e2a9b49.png">

I still have the mypaint brush icons missing and saving issues.

Creating a new scene in the startup popup (Xcode) causes a hang:
<img width="929" alt="Screen Shot 2020-04-05 at 11 37 46 PM" src="https://user-images.githubusercontent.com/21159116/78521052-9f526b80-7796-11ea-8bc4-4bff4d813c17.png">
With Terminal, it's a crash.
[new_scene_crash.txt](https://github.com/opentoonz/opentoonz/files/4435769/new_scene_crash.txt)

**Comment by MrBlockCat** [comment]
That is indeed very strange that what I did works, then! I might try messing with those few lines and seeing which things run on my system. It’s also pretty weird that Terminal works for most people but not me; That’s probably just me messing something up then. And thanks for creating the Xcode build directions! I couldn’t have built anything without them, for whatever’s causing the Terminal build steps to not work for me.

Regardless of all this, though, I’m hoping the GitHub Actions building will help make this work more consistently for everyone!

**Comment by artisteacher** [comment]
I think I've finally found the problem on my end! I had mono framework installed with Unity which is causing my builds to happen with mono's older version of libpng. For now I've removed mono framework. No more crash on making a new scene, mypaint icons are visible, and so far saving has been fine.

Is there a way we can change the build instructions or cmake file so that it uses the correct installation of libpng?

**Comment by MrBlockCat** [comment]
Awesome! If other programs can influence the build, then maybe my issues have to do with me having Eclipse: It's only the Java IDE, I think, but it does install a lot of stuff that could probably mess with OpenToonz builds if it can. Another reason for me to be excited for GitHub Actions, haha

**Comment by MrBlockCat** [comment]
After reinstalling everything on my MacBook, I did some more testing with building, and I'm able to build fine with changing either the QT_PATH or the QT_LIB_PATH lines now (even though QT_LIB_PATH is the wrong line to change). My best guess is that the path QT_PATH points to when it's not touched also worked on my OS, it's just an older version of Qt. (Although I still have the grey arrow issue) Not sure what was causing me trouble before, but all seems well now.

I also just noticed another window besides camera settings that crashes OT when it closes, only on manual builds: Scan & Cleanup > Scan Settings... opens and closes fine in nightly, but crashes when I close the settings menu in the manual build. Here's the error I get:
<img width="746" alt="Screen Shot 2020-04-16 at 10 49 26 PM" src="https://user-images.githubusercontent.com/37186692/79527012-92eece00-8034-11ea-834e-0c258eff1313.png">
in toonz/sources/toonz/scanpopup.cpp
